File: src/data/repo.ts
import Dexie from 'dexie'
import type {
  Download,
  DownloadState,
  Episode,
  EpisodeId,
  FilterMap,
  FilterProfile,
  Podcast,
  PodcastId,
  Progress,
  QueueItem,
  Settings,
  Subscription,
} from '@/core/types'
import { db, type WordOverride } from './db'
import { DEFAULT_PROFILES, DEFAULT_SETTINGS } from './defaults'

/**
 * Rewrites filter maps left in a state nothing will ever advance.
 *
 * Filtering used to be a background queue with 'queued' and 'transcribing' states.
 * Live filtering owns that work in memory now, so any row still carrying those statuses
 * is orphaned — it would sit at "Filtering 0%" forever. Rows that got some analysis done
 * are demoted to 'partial' so the work is not thrown away; empty ones are deleted so the
 * episode simply looks unchecked.
 */
export async function reconcileFilterMaps(): Promise<number> {
  const stale = await db.filterMaps
    .filter((map) => map.status !== 'partial' && map.status !== 'ready' &&
      map.status !== 'failed' && map.status !== 'unsupported')
    .toArray()

  for (const map of stale) {
    if (map.analyzedRanges?.length > 0) {
      await db.filterMaps.put({ ...map, status: 'partial' })
    } else {
      await db.filterMaps.delete(map.episodeId)
    }
  }
  return stale.length
}

/** Seeds built-in profiles and settings. Safe to call on every startup. */
export async function initializeStore(): Promise<void> {
  await db.transaction('rw', db.settings, db.filterProfiles, async () => {
    const existing = await db.settings.get('singleton')
    if (!existing) {
      await db.settings.put({ id: 'singleton', ...DEFAULT_SETTINGS })
    }
    for (const profile of DEFAULT_PROFILES) {
      // put() rather than add() so profile tuning ships in updates, but user-created
      // profiles (which never share these ids) are left alone.
      const current = await db.filterProfiles.get(profile.id)
      if (!current) await db.filterProfiles.put(profile)
    }
  })
}

// --- settings ---------------------------------------------------------------

export async function getSettings(): Promise<Settings> {
  const row = await db.settings.get('singleton')
  return row ?? { ...DEFAULT_SETTINGS }
}

export async function updateSettings(patch: Partial<Settings>): Promise<void> {
  const current = await getSettings()
  await db.settings.put({ id: 'singleton', ...current, ...patch })
}

export async function getActiveProfile(): Promise<FilterProfile> {
  const settings = await getSettings()
  const profile = await db.filterProfiles.get(settings.activeFilterProfileId)
  return profile ?? DEFAULT_PROFILES[1]
}

/** Resolves the profile for an episode, honouring a per-subscription override. */
export async function getProfileForPodcast(podcastId: PodcastId): Promise<FilterProfile> {
  const subscription = await db.subscriptions.get(podcastId)
  if (subscription?.filterProfileId) {
    const profile = await db.filterProfiles.get(subscription.filterProfileId)
    if (profile) return profile
  }
  return getActiveProfile()
}

// --- podcasts and episodes --------------------------------------------------

export async function upsertPodcast(podcast: Podcast): Promise<void> {
  const existing = await db.podcasts.get(podcast.id)
  await db.podcasts.put(existing ? { ...existing, ...podcast } : podcast)
}

export function getPodcast(id: PodcastId): Promise<Podcast | undefined> {
  return db.podcasts.get(id)
}

/**
 * Writes episodes, preserving any fields already stored.
 * Returns the ids that were new, which drives new-episode notifications and
 * auto-download.
 */
export async function upsertEpisodes(episodes: Episode[]): Promise<EpisodeId[]> {
  if (episodes.length === 0) return []
  return db.transaction('rw', db.episodes, async () => {
    const ids = episodes.map((episode) => episode.id)
    const existing = await db.episodes.bulkGet(ids)
    const known = new Set(existing.filter(Boolean).map((episode) => episode!.id))
    await db.episodes.bulkPut(episodes)
    return ids.filter((id) => !known.has(id))
  })
}

export function listEpisodes(podcastId: PodcastId, limit = 200): Promise<Episode[]> {
  return db.episodes
    .where('[podcastId+publishedAt]')
    .between([podcastId, Dexie.minKey], [podcastId, Dexie.maxKey])
    .reverse()
    .limit(limit)
    .toArray()
}

export function getEpisode(id: EpisodeId): Promise<Episode | undefined> {
  return db.episodes.get(id)
}

// --- subscriptions ----------------------------------------------------------

export async function subscribe(podcastId: PodcastId, patch: Partial<Subscription> = {}) {
  const subscription: Subscription = {
    podcastId,
    subscribedAt: Date.now(),
    autoDownload: true,
    autoDownloadLimit: 3,
    notifyOnNew: true,
    ...patch,
  }
  await db.subscriptions.put(subscription)
  return subscription
}

/** Unsubscribes and reclaims storage; listening history is kept. */
export async function unsubscribe(podcastId: PodcastId): Promise<EpisodeId[]> {
  const episodes = await db.episodes.where('podcastId').equals(podcastId).toArray()
  const episodeIds = episodes.map((episode) => episode.id)
  await db.transaction('rw', db.subscriptions, db.downloads, db.filterMaps, async () => {
    await db.subscriptions.delete(podcastId)
    await db.downloads.bulkDelete(episodeIds)
    await db.filterMaps.bulkDelete(episodeIds)
  })
  return episodeIds
}

export function listSubscriptions(): Promise<Subscription[]> {
  return db.subscriptions.orderBy('subscribedAt').reverse().toArray()
}

export function isSubscribed(podcastId: PodcastId): Promise<boolean> {
  return db.subscriptions.get(podcastId).then((row) => row !== undefined)
}

// --- progress ---------------------------------------------------------------

export function getProgress(episodeId: EpisodeId): Promise<Progress | undefined> {
  return db.progress.get(episodeId)
}

export async function saveProgress(
  episodeId: EpisodeId,
  positionSec: number,
  durationSec: number,
  completionThresholdSec: number,
): Promise<void> {
  const nearEnd = durationSec > 0 && positionSec >= durationSec - completionThresholdSec
  const existing = await db.progress.get(episodeId)
  await db.progress.put({
    episodeId,
    positionSec,
    durationSec,
    played: nearEnd || (existing?.played ?? false),
    completedAt: nearEnd ? (existing?.completedAt ?? Date.now()) : existing?.completedAt,
    lastPlayedAt: Date.now(),
  })
}

export async function setPlayed(episodeId: EpisodeId, played: boolean): Promise<void> {
  const existing = await db.progress.get(episodeId)
  await db.progress.put({
    episodeId,
    positionSec: played ? (existing?.durationSec ?? 0) : 0,
    durationSec: existing?.durationSec ?? 0,
    played,
    completedAt: played ? Date.now() : undefined,
    lastPlayedAt: Date.now(),
  })
}

/** Most recently played, unfinished episodes — the "continue listening" row. */
export async function listInProgress(limit = 20): Promise<Progress[]> {
  const rows = await db.progress.orderBy('lastPlayedAt').reverse().limit(limit * 3).toArray()
  return rows.filter((row) => !row.played && row.positionSec > 5).slice(0, limit)
}

// --- downloads --------------------------------------------------------------

export function getDownload(episodeId: EpisodeId): Promise<Download | undefined> {
  return db.downloads.get(episodeId)
}

export async function putDownload(download: Download): Promise<void> {
  await db.downloads.put(download)
}

export async function patchDownload(
  episodeId: EpisodeId,
  patch: Partial<Download>,
): Promise<void> {
  const existing = await db.downloads.get(episodeId)
  if (!existing) return
  await db.downloads.put({ ...existing, ...patch })
}

export function listDownloads(state?: DownloadState): Promise<Download[]> {
  return state ? db.downloads.where('state').equals(state).toArray() : db.downloads.toArray()
}

// --- filter maps ------------------------------------------------------------

export function getFilterMap(episodeId: EpisodeId): Promise<FilterMap | undefined> {
  return db.filterMaps.get(episodeId)
}

export async function putFilterMap(map: FilterMap): Promise<void> {
  await db.filterMaps.put(map)
}

export async function patchFilterMap(
  episodeId: EpisodeId,
  patch: Partial<FilterMap>,
): Promise<void> {
  const existing = await db.filterMaps.get(episodeId)
  if (!existing) return
  await db.filterMaps.put({ ...existing, ...patch })
}

// --- wordlist overrides -----------------------------------------------------

export function listWordOverrides(): Promise<WordOverride[]> {
  return db.wordOverrides.toArray()
}

export async function putWordOverride(override: WordOverride): Promise<void> {
  await db.wordOverrides.put(override)
}

export async function deleteWordOverride(term: string): Promise<void> {
  await db.wordOverrides.delete(term)
}

// --- play queue -------------------------------------------------------------

/**
 * The queue, in play order.
 *
 * Position is dense and rewritten wholesale by the mutations below, so reading is a
 * plain ordered scan and callers never have to reason about gaps.
 */
export function listQueue(): Promise<QueueItem[]> {
  return db.queue.orderBy('position').toArray()
}

/**
 * Adds an episode to the queue.
 *
 * Keyed by episode, so queueing something already queued moves it rather than
 * duplicating it — a queue with the same episode twice is never what was meant, and
 * playing it would strand the second copy at a position already listened to.
 */
export async function enqueueEpisode(
  episodeId: EpisodeId,
  options: { next?: boolean } = {},
): Promise<void> {
  await db.transaction('rw', db.queue, async () => {
    const items = (await db.queue.orderBy('position').toArray()).filter(
      (item) => item.episodeId !== episodeId,
    )
    const entry: QueueItem = { episodeId, position: 0, addedAt: Date.now() }
    const ordered = options.next ? [entry, ...items] : [...items, entry]
    await renumber(ordered)
  })
}

export async function removeFromQueue(episodeId: EpisodeId): Promise<void> {
  await db.transaction('rw', db.queue, async () => {
    await db.queue.delete(episodeId)
    await renumber(await db.queue.orderBy('position').toArray())
  })
}

/**
 * Moves a queued episode by [delta] places, clamped to the ends.
 *
 * Expressed as a relative move rather than an absolute index because that is what the
 * controls do, and it means the caller never has to know the current position — which it
 * would otherwise have to read first, racing anything else editing the queue.
 */
export async function moveInQueue(episodeId: EpisodeId, delta: number): Promise<void> {
  await db.transaction('rw', db.queue, async () => {
    const items = await db.queue.orderBy('position').toArray()
    const from = items.findIndex((item) => item.episodeId === episodeId)
    if (from < 0) return

    const to = Math.min(items.length - 1, Math.max(0, from + delta))
    if (to === from) return

    const [moved] = items.splice(from, 1)
    items.splice(to, 0, moved)
    await renumber(items)
  })
}

/** Takes the next episode off the queue, or undefined when it is empty. */
export async function dequeueNext(): Promise<EpisodeId | undefined> {
  return db.transaction('rw', db.queue, async () => {
    const items = await db.queue.orderBy('position').toArray()
    const next = items.shift()
    if (!next) return undefined
    await db.queue.delete(next.episodeId)
    await renumber(items)
    return next.episodeId
  })
}

export async function clearQueue(): Promise<void> {
  await db.queue.clear()
}

/** Writes back dense positions matching array order. */
async function renumber(items: QueueItem[]): Promise<void> {
  await db.queue.bulkPut(items.map((item, index) => ({ ...item, position: index })))
}

File: kmp/shared/src/commonMain/kotlin/app/filterpod/shared/data/Repo.kt
package app.filterpod.shared.data

import app.cash.sqldelight.coroutines.asFlow
import app.cash.sqldelight.coroutines.mapToList
import app.filterpod.shared.db.FilterPodDb
import app.filterpod.shared.model.Chapter
import app.filterpod.shared.model.Download
import app.filterpod.shared.model.DownloadState
import app.filterpod.shared.model.Episode
import app.filterpod.shared.model.FilterMap
import app.filterpod.shared.model.FilterMapStatus
import app.filterpod.shared.model.FilterProfile
import app.filterpod.shared.model.Podcast
import app.filterpod.shared.model.Progress
import app.filterpod.shared.model.QueueItem
import app.filterpod.shared.model.Settings
import app.filterpod.shared.model.Subscription
import app.filterpod.shared.model.WordOverride
import kotlinx.coroutines.CoroutineDispatcher
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.withContext
import kotlinx.serialization.builtins.ListSerializer
import kotlinx.serialization.json.Json

/**
 * The data layer, ported from src/data/repo.ts with the same semantics.
 *
 * Rows are stored as JSON documents beside their indexed columns, so every rule that
 * was true in Dexie stays true here: dense queue renumbering, the completion-threshold
 * played rule, newer-progress-wins on import, put-if-absent profile seeding.
 */
class Repo(
    private val db: FilterPodDb,
    private val io: CoroutineDispatcher,
) {
    // Lenient + ignoreUnknownKeys: rows written by future versions (or by the
    // TypeScript app, whose optional fields come and go) must never brick a read.
    val json = Json {
        ignoreUnknownKeys = true
        isLenient = true
        encodeDefaults = true
    }

    private val q get() = db.filterPodQueries

    // --- setup ---------------------------------------------------------------

    /** Seeds built-in profiles and settings. Safe to call on every startup. */
    suspend fun initialize() = withContext(io) {
        db.transaction {
            if (q.getKv(KEY_SETTINGS).executeAsOneOrNull() == null) {
                q.putKv(KEY_SETTINGS, json.encodeToString(Settings.serializer(), DEFAULT_SETTINGS))
            }
            for (profile in DEFAULT_PROFILES) {
                // Seed-if-absent: profile tuning ships in updates via migrations, but
                // user-edited rows are left alone.
                if (q.getFilterProfile(profile.id).executeAsOneOrNull() == null) {
                    q.upsertFilterProfile(profile.id, json.encodeToString(FilterProfile.serializer(), profile))
                }
            }
        }
    }

    // --- settings ------------------------------------------------------------

    suspend fun getSettings(): Settings = withContext(io) {
        q.getKv(KEY_SETTINGS).executeAsOneOrNull()
            ?.let { json.decodeFromString(Settings.serializer(), it) }
            ?: DEFAULT_SETTINGS
    }

    suspend fun updateSettings(update: (Settings) -> Settings) = withContext(io) {
        val current = q.getKv(KEY_SETTINGS).executeAsOneOrNull()
            ?.let { json.decodeFromString(Settings.serializer(), it) }
            ?: DEFAULT_SETTINGS
        q.putKv(KEY_SETTINGS, json.encodeToString(Settings.serializer(), update(current)))
    }

    suspend fun getActiveProfile(): FilterProfile {
        val settings = getSettings()
        return getFilterProfile(settings.activeFilterProfileId) ?: DEFAULT_PROFILES[1]
    }

    /** Resolves the profile for an episode, honouring a per-subscription override. */
    suspend fun getProfileForPodcast(podcastId: String): FilterProfile {
        val subscription = getSubscription(podcastId)
        subscription?.filterProfileId?.let { override ->
            getFilterProfile(override)?.let { return it }
        }
        return getActiveProfile()
    }

    // --- podcasts and episodes ----------------------------------------------

    suspend fun upsertPodcast(podcast: Podcast) = withContext(io) {
        val merged = q.getPodcast(podcast.id).executeAsOneOrNull()
            ?.let { json.decodeFromString(Podcast.serializer(), it) }
            ?.let { existing -> mergePodcast(existing, podcast) }
            ?: podcast
        q.upsertPodcast(merged.id, merged.feedUrl, json.encodeToString(Podcast.serializer(), merged))
    }

    suspend fun getPodcast(id: String): Podcast? = withContext(io) {
        q.getPodcast(id).executeAsOneOrNull()?.let { json.decodeFromString(Podcast.serializer(), it) }
    }

    suspend fun listPodcasts(): List<Podcast> = withContext(io) {
        q.listPodcasts().executeAsList().map { json.decodeFromString(Podcast.serializer(), it) }
    }

    /** Writes episodes, returning the ids that were new (drives auto-queue/download). */
    suspend fun upsertEpisodes(episodes: List<Episode>): List<String> = withContext(io) {
        if (episodes.isEmpty()) return@withContext emptyList()
        val fresh = mutableListOf<String>()
        db.transaction {
            for (episode in episodes) {
                val known = q.episodeExists(episode.id).executeAsOne() > 0
                if (!known) fresh += episode.id
                q.upsertEpisode(
                    episode.id, episode.podcastId, episode.publishedAt,
                    json.encodeToString(Episode.serializer(), episode),
                )
            }
        }
        fresh
    }

    suspend fun getEpisode(id: String): Episode? = withContext(io) {
        q.getEpisode(id).executeAsOneOrNull()?.let { json.decodeFromString(Episode.serializer(), it) }
    }

    suspend fun listEpisodes(podcastId: String, limit: Long = 200): List<Episode> = withContext(io) {
        q.listEpisodesForPodcast(podcastId, limit).executeAsList()
            .map { json.decodeFromString(Episode.serializer(), it) }
    }

    fun observeEpisodes(podcastId: String, limit: Long = 200): Flow<List<Episode>> =
        q.listEpisodesForPodcast(podcastId, limit).asFlow().mapToList(io)
            .mapListJson(Episode.serializer())

    /**
     * Records the duration the player actually measured. Feed-declared durations are
     * wrong often enough to matter; measured playback is ground truth once known.
     */
    suspend fun recordMeasuredDuration(episodeId: String, durationSec: Double) = withContext(io) {
        val episode = q.getEpisode(episodeId).executeAsOneOrNull()
            ?.let { json.decodeFromString(Episode.serializer(), it) } ?: return@withContext
        val updated = episode.copy(durationSec = kotlin.math.round(durationSec).toInt())
        q.upsertEpisode(updated.id, updated.podcastId, updated.publishedAt,
            json.encodeToString(Episode.serializer(), updated))
    }

    // --- subscriptions -------------------------------------------------------

    suspend fun subscribe(podcastId: String, base: Subscription? = null, now: Long): Subscription =
        withContext(io) {
            val subscription = base ?: Subscription(podcastId = podcastId, subscribedAt = now)
            q.upsertSubscription(subscription.podcastId, subscription.subscribedAt,
                json.encodeToString(Subscription.serializer(), subscription))
            subscription
        }

    suspend fun putSubscription(subscription: Subscription) = withContext(io) {
        q.upsertSubscription(subscription.podcastId, subscription.subscribedAt,
            json.encodeToString(Subscription.serializer(), subscription))
    }

    suspend fun getSubscription(podcastId: String): Subscription? = withContext(io) {
        q.getSubscription(podcastId).executeAsOneOrNull()
            ?.let { json.decodeFromString(Subscription.serializer(), it) }
    }

    suspend fun listSubscriptions(): List<Subscription> = withContext(io) {
        q.listSubscriptions().executeAsList().map { json.decodeFromString(Subscription.serializer(), it) }
    }

    fun observeSubscriptions(): Flow<List<Subscription>> =
        q.listSubscriptions().asFlow().mapToList(io).mapListJson(Subscription.serializer())

    /** Writes the library's arrangement: dense sortOrder in the order given. */
    suspend fun setLibraryOrder(podcastIds: List<String>) = withContext(io) {
        db.transaction {
            podcastIds.forEachIndexed { index, podcastId ->
                val row = q.getSubscription(podcastId).executeAsOneOrNull() ?: return@forEachIndexed
                val updated = json.decodeFromString(Subscription.serializer(), row).copy(sortOrder = index)
                q.upsertSubscription(updated.podcastId, updated.subscribedAt,
                    json.encodeToString(Subscription.serializer(), updated))
            }
        }
    }

    /** Unsubscribes and reclaims storage; listening history is kept. */
    suspend fun unsubscribe(podcastId: String): List<String> = withContext(io) {
        val episodeIds = q.listEpisodeIdsForPodcast(podcastId).executeAsList()
        db.transaction {
            q.deleteSubscription(podcastId)
            for (id in episodeIds) {
                q.deleteDownload(id)
                q.deleteFilterMap(id)
            }
        }
        episodeIds
    }

    // --- progress ------------------------------------------------------------

    suspend fun getProgress(episodeId: String): Progress? = withContext(io) {
        q.getProgress(episodeId).executeAsOneOrNull()
            ?.let { json.decodeFromString(Progress.serializer(), it) }
    }

    suspend fun saveProgress(
        episodeId: String,
        positionSec: Double,
        durationSec: Double,
        completionThresholdSec: Int,
        now: Long,
    ) = withContext(io) {
        val nearEnd = durationSec > 0 && positionSec >= durationSec - completionThresholdSec
        val existing = q.getProgress(episodeId).executeAsOneOrNull()
            ?.let { json.decodeFromString(Progress.serializer(), it) }
        putProgress(
            Progress(
                episodeId = episodeId,
                positionSec = positionSec,
                durationSec = durationSec,
                played = nearEnd || (existing?.played ?: false),
                completedAt = if (nearEnd) existing?.completedAt ?: now else existing?.completedAt,
                lastPlayedAt = now,
            ),
        )
    }

    suspend fun setPlayed(episodeId: String, played: Boolean, now: Long) = withContext(io) {
        val existing = q.getProgress(episodeId).executeAsOneOrNull()
            ?.let { json.decodeFromString(Progress.serializer(), it) }
        putProgress(
            Progress(
                episodeId = episodeId,
                positionSec = if (played) existing?.durationSec ?: 0.0 else 0.0,
                durationSec = existing?.durationSec ?: 0.0,
                played = played,
                completedAt = if (played) now else null,
                lastPlayedAt = now,
            ),
        )
    }

    suspend fun putProgress(progress: Progress) = withContext(io) {
        q.upsertProgress(progress.episodeId, progress.lastPlayedAt,
            if (progress.played) 1 else 0, json.encodeToString(Progress.serializer(), progress))
    }

    /** Most recently played, unfinished episodes — the "continue listening" row. */
    suspend fun listInProgress(limit: Int = 20): List<Progress> = withContext(io) {
        q.listProgressRecent((limit * 3).toLong()).executeAsList()
            .map { json.decodeFromString(Progress.serializer(), it) }
            .filter { !it.played && it.positionSec > 5 }
            .take(limit)
    }

    fun observeProgress(): Flow<List<Progress>> =
        q.listAllProgress().asFlow().mapToList(io).mapListJson(Progress.serializer())

    // --- downloads -----------------------------------------------------------

    suspend fun getDownload(episodeId: String): Download? = withContext(io) {
        q.getDownload(episodeId).executeAsOneOrNull()
            ?.let { json.decodeFromString(Download.serializer(), it) }
    }

    suspend fun putDownload(download: Download) = withContext(io) {
        q.upsertDownload(download.episodeId, download.state.name, download.startedAt,
            json.encodeToString(Download.serializer(), download))
    }

    suspend fun patchDownload(episodeId: String, patch: (Download) -> Download) = withContext(io) {
        val existing = q.getDownload(episodeId).executeAsOneOrNull()
            ?.let { json.decodeFromString(Download.serializer(), it) } ?: return@withContext
        val updated = patch(existing)
        q.upsertDownload(updated.episodeId, updated.state.name, updated.startedAt,
            json.encodeToString(Download.serializer(), updated))
    }

    suspend fun listDownloads(state: DownloadState? = null): List<Download> = withContext(io) {
        val rows = if (state != null) q.listDownloadsByState(state.name).executeAsList()
        else q.listDownloads().executeAsList()
        rows.map { json.decodeFromString(Download.serializer(), it) }
    }

    fun observeDownloads(): Flow<List<Download>> =
        q.listDownloads().asFlow().mapToList(io).mapListJson(Download.serializer())

    // --- filter maps ---------------------------------------------------------

    suspend fun getFilterMap(episodeId: String): FilterMap? = withContext(io) {
        q.getFilterMap(episodeId).executeAsOneOrNull()
            ?.let { json.decodeFromString(FilterMap.serializer(), it) }
    }

    suspend fun putFilterMap(map: FilterMap) = withContext(io) {
        q.upsertFilterMap(map.episodeId, map.status.name,
            json.encodeToString(FilterMap.serializer(), map))
    }

    /**
     * The stored map only when it is a valid head start for these rules: same engine,
     * same wordlist, same profile, not failed. Anything else vouches for audio these
     * rules never examined — discarded, exactly as the TypeScript drivers do.
     */
    suspend fun getValidFilterMap(episodeId: String, profileId: String): FilterMap? {
        val stored = getFilterMap(episodeId) ?: return null
        val valid = stored.status != FilterMapStatus.FAILED &&
            stored.engineVersion == ENGINE_VERSION &&
            stored.wordlistVersion == WORDLIST_VERSION &&
            stored.profileId == profileId
        return if (valid) stored else null
    }

    // --- word overrides ------------------------------------------------------

    suspend fun listWordOverrides(): List<WordOverride> = withContext(io) {
        q.listWordOverrides().executeAsList().map { json.decodeFromString(WordOverride.serializer(), it) }
    }

    suspend fun putWordOverride(override: WordOverride) = withContext(io) {
        q.upsertWordOverride(override.term, json.encodeToString(WordOverride.serializer(), override))
    }

    suspend fun deleteWordOverride(term: String) = withContext(io) { q.deleteWordOverride(term) }

    // --- filter profiles -----------------------------------------------------

    suspend fun getFilterProfile(id: String): FilterProfile? = withContext(io) {
        q.getFilterProfile(id).executeAsOneOrNull()
            ?.let { json.decodeFromString(FilterProfile.serializer(), it) }
    }

    suspend fun listFilterProfiles(): List<FilterProfile> = withContext(io) {
        q.listFilterProfiles().executeAsList().map { json.decodeFromString(FilterProfile.serializer(), it) }
    }

    suspend fun putFilterProfile(profile: FilterProfile) = withContext(io) {
        q.upsertFilterProfile(profile.id, json.encodeToString(FilterProfile.serializer(), profile))
    }

    // --- chapters ------------------------------------------------------------

    suspend fun getCachedChapters(episodeId: String): List<Chapter>? = withContext(io) {
        q.getChapterCache(episodeId).executeAsOneOrNull()
            ?.let { json.decodeFromString(ListSerializer(Chapter.serializer()), it) }
    }

    suspend fun putCachedChapters(episodeId: String, chapters: List<Chapter>, now: Long) =
        withContext(io) {
            q.upsertChapterCache(episodeId, now,
                json.encodeToString(ListSerializer(Chapter.serializer()), chapters))
        }

    // --- play queue ----------------------------------------------------------

    suspend fun listQueue(): List<QueueItem> = withContext(io) {
        q.listQueue().executeAsList().map { QueueItem(it.episodeId, it.position.toInt(), it.addedAt) }
    }

    fun observeQueue(): Flow<List<QueueItem>> =
        q.listQueue().asFlow().mapToList(io).let { flow ->
            kotlinx.coroutines.flow.flow {
                flow.collect { rows ->
                    emit(rows.map { QueueItem(it.episodeId, it.position.toInt(), it.addedAt) })
                }
            }
        }

    /**
     * Adds an episode to the queue. Keyed by episode, so queueing something already
     * queued moves it rather than duplicating it.
     */
    suspend fun enqueueEpisode(episodeId: String, next: Boolean = false, now: Long) = withContext(io) {
        db.transaction {
            val items = q.listQueue().executeAsList()
                .map { QueueItem(it.episodeId, it.position.toInt(), it.addedAt) }
                .filter { it.episodeId != episodeId }
            val entry = QueueItem(episodeId, 0, now)
            renumber(if (next) listOf(entry) + items else items + entry)
        }
    }

    suspend fun removeFromQueue(episodeId: String) = withContext(io) {
        db.transaction {
            q.deleteQueueItem(episodeId)
            renumber(q.listQueue().executeAsList().map { QueueItem(it.episodeId, it.position.toInt(), it.addedAt) })
        }
    }

    /** Rewrites the queue to exactly this order. Drag-reorder writes the whole list. */
    suspend fun reorderQueue(episodeIds: List<String>) = withContext(io) {
        db.transaction {
            val byId = q.listQueue().executeAsList()
                .associateBy({ it.episodeId }, { QueueItem(it.episodeId, it.position.toInt(), it.addedAt) })
            renumber(episodeIds.mapNotNull { byId[it] })
        }
    }

    /** Position 0 is whatever is playing; advancing = drop the finished head, read again. */
    suspend fun headOfQueue(): String? = withContext(io) {
        q.listQueue().executeAsList().firstOrNull()?.episodeId
    }

    private fun renumber(items: List<QueueItem>) {
        items.forEachIndexed { index, item ->
            q.upsertQueueItem(item.episodeId, index.toLong(), item.addedAt)
        }
    }

    // --- kv ------------------------------------------------------------------

    suspend fun getKv(key: String): String? = withContext(io) { q.getKv(key).executeAsOneOrNull() }
    suspend fun putKv(key: String, value: String) = withContext(io) { q.putKv(key, value) }

    private fun mergePodcast(existing: Podcast, incoming: Podcast): Podcast = incoming.copy(
        // Refresh must not erase validators or history the incoming row lacks.
        etag = incoming.etag ?: existing.etag,
        lastModified = incoming.lastModified ?: existing.lastModified,
        lastFetchedAt = incoming.lastFetchedAt ?: existing.lastFetchedAt,
        lastSuccessAt = incoming.lastSuccessAt ?: existing.lastSuccessAt,
        lastFetchError = incoming.lastFetchError,
    )

    private fun <T> Flow<List<String>>.mapListJson(
        serializer: kotlinx.serialization.KSerializer<T>,
    ): Flow<List<T>> = kotlinx.coroutines.flow.flow {
        collect { rows -> emit(rows.map { json.decodeFromString(serializer, it) }) }
    }

    companion object {
        const val KEY_SETTINGS = "settings"
        /** Stamp written once the Capacitor app's state has been imported. */
        const val KEY_IMPORTED_AT = "importedAt"
    }
}
